Makefile.cache

```diff
@@ -0,0 +1,671 @@
+# for that module.
+#
+# It provides two levels of caching.
+#     Global cache := Module doesnt have any local changes.
+#     Local cache  := Used for target when one of its dependency file is modified.
+#
+#
+# Steps for adding new module is given as a template.
+# Template File : rules/template.dep
+#
+
+
+# Common files and FLAGS
+# Run the 'touch nocommon' command in the base directory to exclude the common files from caching
+SONIC_COMMON_FILES_LIST :=  $(if $(wildcard nocommon),, .platform slave.mk rules/functions Makefile.cache)
+SONIC_COMMON_FLAGS_LIST :=  $(CONFIGURED_PLATFORM) \
+                            $(SONIC_DPKG_CACHE_SOURCE) $(SONIC_DEBUGGING_ON) \
+                            $(SONIC_PROFILING_ON) $(SONIC_SANITIZER_ON)
+SONIC_COMMON_DPKG_LIST  :=  debian/control debian/changelog debian/rules \
+                            debian/compat debian/install debian/copyright
+SONIC_COMMON_BASE_FILES_LIST  := sonic-slave-jessie/Dockerfile.j2 sonic-slave-jessie/Dockerfile.user \
+                                 sonic-slave-stretch/Dockerfile.j2 sonic-slave-stretch/Dockerfile.user
+
+
+
+include $(RULES_PATH)/*.dep
+
+
+ifneq ($(CONFIGURED_PLATFORM), undefined)
+ifeq ($(PDDF_SUPPORT), y)
+include $(PLATFORM_PDDF_PATH)/rules.dep
+endif
+
+-include $(PLATFORM_PATH)/rules.dep
+endif
+
+$(info "SONIC_DPKG_CACHE_METHOD"         : "$(SONIC_DPKG_CACHE_METHOD)")
+ifneq ($(SONIC_DPKG_CACHE_METHOD),none)
+$(info "DPKG_CACHE_PATH"                 : "$(SONIC_DPKG_CACHE_SOURCE)")
+endif
+$(info )
+
+
+###############################################################################
+## Canned sequences
+###############################################################################
+
+SONIC_DPKG_CACHE_DIR    := /dpkg_cache
+MOD_CACHE_LOCK_SUFFIX   := cache_access
+MOD_CACHE_LOCK_TIMEOUT  := 3600
+SONIC_DPKG_LOCAL_CACHE_DIR=${TARGET_PATH}/cache
+$(shell test -d $(SONIC_DPKG_LOCAL_CACHE_DIR) || \
+    mkdir -p $(SONIC_DPKG_LOCAL_CACHE_DIR) && chmod 777 $(SONIC_DPKG_LOCAL_CACHE_DIR) )
+$(shell test -w $(SONIC_DPKG_CACHE_DIR) || sudo chmod 777 $(SONIC_DPKG_CACHE_DIR) ) 
+
+DOCKER_LOCKFILE_SUFFIX  := access
+DOCKER_LOCKFILE_TIMEOUT := 1200
+
+# Lock macro for shared file access
+# Lock is implemented through flock command with a specified timeout value
+# Lock file is created in the specified directory, a separate one for each target file name
+# A designated suffix is appended to each target file name, followed by .lock
+#
+# Parameters:
+#  $(1) - target file name (without path)
+#  $(2) - lock file path (only)
+#  $(3) - designated lock file suffix
+#  $(4) - flock timeout (in seconds)
+#
+# $(call MOD_LOCK,file,path,suffix,timeout)
+define MOD_LOCK
+	if [[ ! -f $(2)/$(1)_$(3).lock ]]; then
+		touch $(2)/$(1)_$(3).lock
+		chmod 777 $(2)/$(1)_$(3).lock;
+	fi
+	$(eval $(1)_lock_fd=$(subst ~,_,$(subst -,_,$(subst +,_,$(subst .,_,$(1))))))
+	exec {$($(1)_lock_fd)}<"$(2)/$(1)_$(3).lock";
```

Comment:
Do you know what is different between the following two commands?
1) exec {lock}<2.lock
2) exec {lock}>2.lock  or exec {lock}<>2.lock
The first one does not work when using NFS, the second one works as expected. Can we change to use as the second one?

root@SonicVM2:/nfs/dpkg_cache4# ll
total 8
drwxrwxr-- 2 nobody 4294967294 4096 Mar  6 13:10 ./
drwxrwx--- 7   1002       1003 4096 Mar  5 01:47 ../
root@SonicVM2:/nfs/dpkg_cache4# touch 2.lock
root@SonicVM2:/nfs/dpkg_cache4# exec {lock}<2.lock
root@SonicVM2:/nfs/dpkg_cache4# flock -x -w 10 $lock
**flock: 11: Bad file descriptor**
root@SonicVM2:/nfs/dpkg_cache4# exec {lock}<&-
root@SonicVM2:/nfs/dpkg_cache4# exec {lock}>2.lock  
root@SonicVM2:/nfs/dpkg_cache4# flock -x -w 10 $lock
root@SonicVM2:/nfs/dpkg_cache4# echo $?
0
root@SonicVM2:/nfs/dpkg_cache4# exec {lock}<&-
root@SonicVM2:/nfs/dpkg_cache4#

Reply:
Fixed